### ISSUE TITLE
[FBX] Vertex colors need vertex duplication

### ISSUE THREAD

**Issue by RevoluPowered** [godotengine/godot#44297]
I need to fix vertex colors in the importer, it's not doing the duplication step it's collecting first and averaging color currently.

Issue was reported via email to myself.

**Godot version:**
3.2.3 beta 3

**OS/device including version:**
Windows, Linux, Mac

**Issue description:**
vertex colors look incorrect

**Steps to reproduce:**
1. open something with vertex colors 

**Steps to fix**
- enable duplication step for colors too.

**Comment by RevoluPowered** [comment]
File high level data

```c
[doc] fbx header: Kaydara FBX Binary  
[doc] is binary
FBX Version: 7400
FBX App Name: 3ds Max
FBX App Vendor: Autodesk
FBX App Version: 2021
FBX unit scale import value: 2.54
FBX unit scale is: 0.0254
mesh name: Geometry::
Mesh statistics 
uv_0: indexes: 0 data: 0
uv_1: indexes: 0 data: 0
vertices: 6858
debugging node name: name: V2Grp001
raw rotation: 0, 0, 0
raw pre_rotation 0, 0, 0
raw post_rotation 0, 0, 0
Model: Model::V2Grp001 Has inherit type: 1
valid pre_rotation: -90, 0, 0 euler conversion: -90, 0, 0
valid post_rotation: 0, 0, 0 euler conversion: -90, 0, 0
local xform calculation 0, 0, 0 rotation: 0.00002, 0, 0
scale of node: 1, 1, 1
---------------------------------------------------------------
global xform:  0, 0, 0 rotation: 0.00002, 0, 0
[doc] new node V2Grp001
debugging node name: name: VDLt001
raw rotation: 0, 0, 0
raw pre_rotation 0, 0, 0
raw post_rotation 0, 0, 0
Model: Model::VDLt001 Has inherit type: 1
valid pre_rotation: 0, 0, 0 euler conversion: -0, 0, 0
valid post_rotation: -90, 0, 0 euler conversion: -0, 0, 0
local xform calculation 0, 0, 0 rotation: -179.999985, 0, 0
scale of node: 1, 1, 1
---------------------------------------------------------------
global xform:  0, 0, 0 rotation: 0.000041, 0, 0
[doc] new node VDLt001
debugging node name: name: V2Xfm001
raw rotation: 0, 0, 0
raw pre_rotation 0, 0, 0
raw post_rotation 0, 0, 0
Model: Model::V2Xfm001 Has inherit type: 1
valid pre_rotation: 0, 0, 0 euler conversion: -0, 0, 0
valid post_rotation: 0, 0, 0 euler conversion: -0, 0, 0
local xform calculation 0, 0, 0 rotation: -0, 0, 0
scale of node: 1, 1, 1
---------------------------------------------------------------
global xform:  0, 0, 0 rotation: 0.00002, 0, 0
[doc] new node V2Xfm001
debugging node name: name: VIFS001
raw rotation: 0, 0, 0
raw pre_rotation 0, 0, 0
raw post_rotation 0, 0, 0
Model: Model::VIFS001 Has inherit type: 1
valid pre_rotation: 0, 0, 0 euler conversion: -0, 0, 0
valid post_rotation: 0, 0, 0 euler conversion: -0, 0, 0
local xform calculation 0, 0, 0 rotation: -0, 0, 0
scale of node: 1, 1, 1
---------------------------------------------------------------
global xform:  0, 0, 0 rotation: 0.00002, 0, 0
[doc] new node VIFS001
[material] material name: Material #25
FBX Material parameter: 3dsMax
FBX Material parameter: 3dsMax|ClassIDa
FBX Material parameter: 3dsMax|ClassIDb
FBX Material parameter: 3dsMax|ORIGINAL_MTL
FBX Material parameter: 3dsMax|Parameters
FBX Material parameter: 3dsMax|Parameters|aniso_angle_map
FBX Material parameter: 3dsMax|Parameters|aniso_angle_map_on
FBX Material parameter: 3dsMax|Parameters|aniso_channel
FBX Material parameter: 3dsMax|Parameters|aniso_mode
FBX Material parameter: 3dsMax|Parameters|anisoangle
FBX Material parameter: 3dsMax|Parameters|anisotropy
FBX Material parameter: 3dsMax|Parameters|anisotropy_map
FBX Material parameter: 3dsMax|Parameters|anisotropy_map_on
FBX Material parameter: 3dsMax|Parameters|base_color
FBX Material parameter: 3dsMax|Parameters|base_color_map_on
FBX Material parameter: 3dsMax|Parameters|base_weight
FBX Material parameter: 3dsMax|Parameters|base_weight_map
FBX Material parameter: 3dsMax|Parameters|base_weight_map_on
FBX Material parameter: 3dsMax|Parameters|brdf_curve
FBX Material parameter: 3dsMax|Parameters|brdf_high
FBX Material parameter: 3dsMax|Parameters|brdf_low
FBX Material parameter: 3dsMax|Parameters|brdf_mode
FBX Material parameter: 3dsMax|Parameters|coat_affect_color
FBX Material parameter: 3dsMax|Parameters|coat_affect_roughness
FBX Material parameter: 3dsMax|Parameters|coat_color
FBX Material parameter: 3dsMax|Parameters|coat_color_map
FBX Material parameter: 3dsMax|Parameters|coat_color_map_on
FBX Material parameter: 3dsMax|Parameters|coat_ior
FBX Material parameter: 3dsMax|Parameters|coat_map
FBX Material parameter: 3dsMax|Parameters|coat_map_on
FBX Material parameter: 3dsMax|Parameters|coat_rough_map
FBX Material parameter: 3dsMax|Parameters|coat_rough_map_on
FBX Material parameter: 3dsMax|Parameters|coat_roughness
FBX Material parameter: 3dsMax|Parameters|coat_roughness_inv
FBX Material parameter: 3dsMax|Parameters|coating
FBX Material parameter: 3dsMax|Parameters|cutout_map
FBX Material parameter: 3dsMax|Parameters|cutout_map_on
FBX Material parameter: 3dsMax|Parameters|diff_rough_map
FBX Material parameter: 3dsMax|Parameters|diff_rough_map_on
FBX Material parameter: 3dsMax|Parameters|diff_roughness
FBX Material parameter: 3dsMax|Parameters|displacement_map
FBX Material parameter: 3dsMax|Parameters|displacement_map_amt
FBX Material parameter: 3dsMax|Parameters|displacement_map_on
FBX Material parameter: 3dsMax|Parameters|emission
FBX Material parameter: 3dsMax|Parameters|emission_map_on
FBX Material parameter: 3dsMax|Parameters|emit_color
FBX Material parameter: 3dsMax|Parameters|emit_color_map
FBX Material parameter: 3dsMax|Parameters|emit_color_map_on
FBX Material parameter: 3dsMax|Parameters|emit_kelvin
FBX Material parameter: 3dsMax|Parameters|emit_luminance
FBX Material parameter: 3dsMax|Parameters|material_mode
FBX Material parameter: 3dsMax|Parameters|metalness
FBX Material parameter: 3dsMax|Parameters|metalness_map_on
FBX Material parameter: 3dsMax|Parameters|refl_color
FBX Material parameter: 3dsMax|Parameters|refl_color_map
FBX Material parameter: 3dsMax|Parameters|refl_color_map_on
FBX Material parameter: 3dsMax|Parameters|reflectivity
FBX Material parameter: 3dsMax|Parameters|reflectivity_map
FBX Material parameter: 3dsMax|Parameters|reflectivity_map_on
FBX Material parameter: 3dsMax|Parameters|roughness
FBX Material parameter: 3dsMax|Parameters|roughness_inv
FBX Material parameter: 3dsMax|Parameters|roughness_map_on
FBX Material parameter: 3dsMax|Parameters|scattering
FBX Material parameter: 3dsMax|Parameters|scattering_map
FBX Material parameter: 3dsMax|Parameters|scattering_map_on
FBX Material parameter: 3dsMax|Parameters|sss_color
FBX Material parameter: 3dsMax|Parameters|sss_color_map
FBX Material parameter: 3dsMax|Parameters|sss_color_map_on
FBX Material parameter: 3dsMax|Parameters|sss_depth
FBX Material parameter: 3dsMax|Parameters|sss_scale
FBX Material parameter: 3dsMax|Parameters|sss_scale_map
FBX Material parameter: 3dsMax|Parameters|sss_scale_map_on
FBX Material parameter: 3dsMax|Parameters|sss_scatter_color
FBX Material parameter: 3dsMax|Parameters|thin_walled
FBX Material parameter: 3dsMax|Parameters|trans_color
FBX Material parameter: 3dsMax|Parameters|trans_color_map
FBX Material parameter: 3dsMax|Parameters|trans_color_map_on
FBX Material parameter: 3dsMax|Parameters|trans_depth
FBX Material parameter: 3dsMax|Parameters|trans_ior
FBX Material parameter: 3dsMax|Parameters|trans_ior_map
FBX Material parameter: 3dsMax|Parameters|trans_ior_map_on
FBX Material parameter: 3dsMax|Parameters|trans_rough_map
FBX Material parameter: 3dsMax|Parameters|trans_rough_map_on
FBX Material parameter: 3dsMax|Parameters|trans_roughness
FBX Material parameter: 3dsMax|Parameters|trans_roughness_inv
FBX Material parameter: 3dsMax|Parameters|trans_roughness_lock
FBX Material parameter: 3dsMax|Parameters|transparency
FBX Material parameter: 3dsMax|Parameters|transparency_map
FBX Material parameter: 3dsMax|Parameters|transparency_map_on
FBX Material parameter: 3dsMax|SuperClassID
FBX Material parameter: DiffuseColor
FBX Material parameter: EmissiveColor
FBX Material parameter: ShadingModel
FBX Material parameter: TransparencyFactor
[doc] Skeleton Bone count: 0
[doc] [2100063860208] mesh: VIFS001
 FBX file is missing indexing array
 scene/resources/surface_tool.cpp:927 - Condition "!(format & Mesh::ARRAY_FORMAT_TEX_UV)" is true.
[doc] Skeleton Bone count: 0
Valid animation stack has been found: Take 001
r_start 0, r_stop 3.333333
start_time0 end_time 3.333333
anim duration : 3.333333
Creating animation player
Animation length: 3.333333 seconds
FBX Animation layers: 1
Layer: AnimLayerBaseLayer,  AnimCurveNode count 5
track defaults: p(0, 0, 0) s(1, 1, 1) r(0, 0, 0, 1)
track defaults: p(0, 0, 0) s(1, 1, 1) r(0, 0, 0, 1)
track defaults: p(0, 0, 0) s(1, 1, 1) r(0, 0, 0, 1)
track defaults: p(0, 0, 0) s(1, 1, 1) r(0, 0, 0, 1)
Loading resource: res://Vertex colour/Material #25.material
Saving scene to: res://.import/Kragarm_I_Traeger_Streckenlast_01_sig11.fbx-d9a4bbb6714aabcf22ff98bc190aa622.scn```

**Comment by RevoluPowered** [comment]
First fix:
- 
```

	ERR_FAIL_COND_V_MSG(p_mapping_data.ref_type == FBXDocParser::MeshGeometry::ReferenceType::index_to_direct && p_mapping_data.index.size() == 0, (HashMap<int, R>()), "FBX file is missing indexing array");
	ERR_FAIL_COND_V_MSG(p_mapping_data.ref_type == FBXDocParser::MeshGeometry::ReferenceType::index && p_mapping_data.index.size() == 0, (HashMap<int, R>()), "The FBX seems corrupted");
```
1. this needs remapping to either array.
2. reorganize_vertices() needs to duplicate vert colors like the rest of the items do.

**Comment by RevoluPowered** [comment]
checked index is broken entirely in the file need to dig into ascii file

**Comment by akien-mga** [comment]
Fixed by #44305.
